### ISSUE TITLE
Rethinking popups

### ISSUE THREAD

**Issue by hoelzro** [Jermolene/TiddlyWiki5#3591]
[Original post on Google Groups - TiddlyWikiDev](https://groups.google.com/forum/#!topic/tiddlywikidev/TgJ823FHlXw)

Hi everyone,

I have been thinking for a while about the popup system after trying to fix a bug I reported, and I think the popup system could benefit from a little overhaul.  I hope I'm not overreaching here - I would just really appreciate some feedback on these ideas!

# What circumstances are the popup system used in?

To start, I looked through all references to "popup" in the TiddlyWiki codebase (not including non-official plugins).  I found the plugin system is used for these four UI patterns:

-  Dropdowns (ex. search results, edit toolbar popups) - this is probably the most common
-  Menus (a special case of dropdown - ex. the export menu)
-  Accordion UI pattern (ex. `$:/Manager`, plugin info, tiddler info when you click "more > info" in the view toolbar)
-  Dialogs (ex. the confirmation prompt when you try to delete tiddlers matching a filter in `$:/AdvancedSearch`)

So, bearing those four use cases in mind...

# What are the requirements of the popup system?

## We need popups to open when we activate the corresponding UI element

This mostly corresponds to `<$edit-text>` and `<$button>` - but I don't see why we should limit ourselves to these.  I like keeping the current shorthand - it makes the common case easy to do, plus it's backwards compatible.

## We need popups to close under certain conditions

In my opinion, a popup should close under two conditions:

-  If I click on a UI element that completes the workflow that the popup was summoned to complete (applies to dropdowns, menus, and dialogs).  For example:
    - Clicking on a tag in the tag dropdown
    - Clicking on a button in a "confirm" dialog popup
    - Clicking on a tiddler in the "create wikitext link" dropdown

- If I click on or _otherwise grant focus_ to an element outside of the popup
    - Dialogs are kind of the exception here - clicking outside of a dialog should probably not close it, but I don't think it's the end of the world if it does
    - Another exception is the "popup handle" - an input element that's bound to the popup.  A prime example of this is the search query edit field - clicking in it shouldn't close the search results

Clicking within a popup shouldn't close it, I think - which is contrary to the current default behavior.  I think most people would expect a popup to stick around until they interact with it meaningfully, dimiss it with the button they summoned it with, or click outside of it.

# What would a new popup system look like?

I rather like the idea of a new action widget - `<$action-popup>` - to be the common interface between wikitext and the popup mechanism.  It can be the plumbing upon which the higher-level interfaces offered by `<$edit-text>` and `<$button>` are built.

A widget can be bound to a popup (similar to how `focusPopup` works now) - clicking on a widget will not close its own popup (but will close others).

The current popup system currently maintains state in JS; I would really like for the state to be maintained entirely in the tiddler store for consistency purposes.

The current system supports nesting of popups - as far as I can tell, this is only really used for popups that behave like menus, so maybe the right thing to do is to break out a distinct "menu" widget.  Doing so might simplify state management quite a bit!  It might also make sense to break out a distinct "dialog" widget.

I'm tempted to tie certain bits of extra state to a popup - eg. the search text for "create wikitext link", so that it gets cleared as well when a popup is closed, but maybe that's trying to do too much.

# Thinking Ahead

There are some other things that users might want popups to do someday; for example, Tobias' appear plugin augments the popup system with some additional behavior, and users might want features like tearoff menus in the future.  We can't predict everything everyone might want, but it might be a good time to incorporate augmentations like Tobias' and identify other possible uses for the popup system.

Thanks for reading!  Please let me know if you have any suggestions, or if you find flaws in my ideas!  I'd be happy to undertake this work, as long as it's a solution everyone's happy with.

**Comment by hoelzro** [comment]
Simon on the Google Group brought up a good point:

> The popup mechanism should also somehow allow access to the tiddler the popup is bound to if it's a tiddler in the story river. Your proposal of storing popup states in tiddlers would allow that

This brings to light an interesting behavior - open a popup in a tiddler, close that tiddler, re-open that tiddler.  I think most users would expect the popup to _not_ be present upon re-open.  Clicking the "close" button, following the rules above, should close the popup, but what if the user uses a keyboard shortcut to close the tiddler?  I would argue that perhaps the close rules to expand to include any interaction with a UI element outside of the popup (including activating a UI element via keyboard shortcut).

That would solve the "open - close - re-open" issue, but it feels like behavior that _happens_ to solve it.

**Comment by Jermolene** [comment]
Hi @hoelzro great point. I think you're right: keyboard shortcuts should cancel outstanding popups.

**Comment by AnthonyMuscio** [comment]
Please proceed down this path. Popups are underutilised in TiddlyWiki in part because the are cumbersome and not expanded on very much in documentation.

**Comment by BurningTreeC** [comment]
Hi all, I believe the `navigator` widget could handle a `tm-popup` message. There we can access story information ...

The triggering widget of a popup should be passed to the popup handler, so that he can extract information about the popup from it

So should do the reveal widget which is the one that shows / hides a popup in the end

I think with both informations from the triggering and showing widget within the navigator, we have access to everything we need

**Comment by pmario** [comment]
Good thoughts. ... 
TLDR;

It's OK to improve traditional popup-handling. BUT it needs to be simple to eliminate them!!

--------------

At the moment, popups are a "special" elements, created with the reveal-widget. So for me they are part of the UI, that have special CSS rules, to let them stand out visually. ... but it's relatively easy to include them as part of the "story based" UI

I personally don't want to have popups and modals at all. I try to completely remove them for my themes. 

eg: 

 - open https://wikilabs.github.io/#advanced-search-plus
 - click the Search button
 - click the "plugin" tag-pill at the tiddler

**Comment by AnthonyMuscio** [comment]
Folks, As BTC mentioned the the navigator widget, I am wondering if popups are arguably related to actions that we may apply to tiddlers?, As previously mentioned by me I would like the opening (navigator Widget) closure (close and other buttons) to trigger an action specified in the field of a given tiddler (or other conditions like on open if tagged). This would allow a set of actions to be associated with opening/closing tiddlers. One of these can be popups, for example a popup asking to to select the project for a tiddler if no project is set, another could start a timer which, ends and computes the difference on closure of the tiddler.

I am not asking for a specific solution here, only asking that we can provide the infrastructure to increase the design options for non coders. I think revisiting the popup mechanisium is the idea opportunity to do this.

**Comment by hoelzro** [comment]
@BurningTreeC I like the idea of having the navigator handle things!

@pmario Thanks for your perspective - I hadn't considered the idea of TW sans popups, but I think it would be good to think about this regarding the redesign.  Interestingly enough, your implementation of the tag pill falls under the "accordion" functionality I mentioned above - maybe accordion/modal could be different "styles" used to implement popup functionality?

@AnthonyMuscio Interesting - I think this lends more support to the idea of an `action-popup` action widget.  That would compose very well with "activate these actions upon open" , without needing to explicitly encode that logic into the popup mechanism itself.

**Comment by BurningTreeC** [comment]
@hoelzro , the `reveal` widget seems to be the central part

it knows when something opens, it's always there where the popup is, so from within it we have access to the dom and widget tree, so we can find parent/sibling widgets with open popups. Maybe each reveal widget with a popup can register itself somewhere when it executes, so that those reveal widgets that create a connected structure can already be marked as connected

**Comment by telmiger** [comment]
Interesting thoughts, I would like to add mine. I tend to invent my own popups to circumvent the standard way which always felt a bit complicated, while on the other hand adding an element (to the DOM) and showing/hiding it via state tiddlers and CSS seemed more intuitive ...

The latest example is my plugin Simple Search (see https://tid.li/tw5/plugins.html) that solves at least two issues that bothered me: 
- closing: my close button covers the whole screen, just beneeth the popup area
- add to the story: like Advanced Search, I think Simple Search is a step in the user journey and should be recorded in $:/HistoryList – so I open the tiddler at the top of the story list (like any other tiddler) but with height 0 in the story and popping it’s content full-screen on a high z-index.

Covering the whole screen also solves another issue I often had with popups: It is hard (and seems sometimes impossible) to position them easily accessible, in context of the button clicked and fully inside the viewport on every screen, especially on mobile.

Does this add another case to the list of the OP?
1. Dropdown/Menu/Chooser
2. Accordion/Reveal/Details
3. Dialog
4. Tiddler

Is it helpful to draw a line between 1/2 as parts of an in-tiddler UI and 3/4 as presenting a new information layer, putting everything else in the background?

**Comment by hoelzro** [comment]
@telmiger Thanks for your input!  It's interesting that you mentioned "tiddler" as use case 4 - I was thinking about the idea of "extracting" a popup to a tiddler in the story list (I actually have a tweak to my own wiki to extract a search to a standalone tiddler in the story).  This also reminds me of @pmario's feedback.  I'm also reminded of [the enhanced search plugin](http://spangenhelm.github.io/TW5-EnhancedSearchResults/).

I'm wondering if popups should be able to flag themselves as "storyable" - ex. the search results could easy be presented in the story river, whereas that doesn't really make sense for the "export tiddler" menu.  Then again, maybe "menus" need to be their own distinct entity going forward.

Regarding the close button - I want to make sure an improved system ships (preferably in 5.1.19), so I'm a little concerned in getting mired in accommodating all possible styles - that being said, I'd like to really consider all of the styles people like you and @pmario use to grant people flexibility if possible!

:point_up: regarding that - I'm _really_ bad at UI/design, so if anyone has some thoughts on how to fit everyone's use cases into the new system, please chime in!

I plan on updating my original draft with the feedback within the next few days.
